src/sync/remote_mongo_collection.cpp

```diff
@@ -21,461 +21,411 @@
+                                                      util::Optional<AppError>)> completion_block)
 {
-    if (value && !error) {
-        
-        try {
-            auto json = nlohmann::json::parse(*value);
-            auto matched_count_string = json.at("matchedCount").at("$numberInt").get<std::string>();
-            auto matched_count = std::stoull(matched_count_string);
-            auto modified_count_string = json.at("modifiedCount").at("$numberInt").get<std::string>();
-            auto modified_count = std::stoull(modified_count_string);
-            std::string upserted_id;
-            auto it = json.find("upsertedId");
-            if (it != json.end()) {
-                upserted_id = json.at("upsertedId").at("$oid").get<std::string>();
-            }
-            
-            return completion_block(RemoteMongoCollection::RemoteUpdateResult {
-                matched_count,
-                modified_count,
-                upserted_id
-            }, error);
-        } catch (const std::exception& e) {
-            return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
-        }
+    
+    if (error) {
+        return completion_block({}, error);
     }
     
-    return completion_block({}, error);
+    try {
+        auto document = static_cast<bson::BsonDocument>(*value);
+        auto matched_count = static_cast<int32_t>(document["matchedCount"]);
+        auto modified_count = static_cast<int32_t>(document["modifiedCount"]);
+
+        Optional<ObjectId> upserted_id;
+        auto it = document.find("upsertedId");
+        if (it != document.end()) {
+            upserted_id = static_cast<ObjectId>(document["upsertedId"]);
+        }
+        
+        return completion_block(RemoteMongoCollection::RemoteUpdateResult {
+            matched_count,
+            modified_count,
+            upserted_id
+        }, error);
+    } catch (const std::exception& e) {
+        return completion_block({}, AppError(make_error_code(JSONErrorCode::bad_bson_parse), e.what()));
+    }
 }
 
-void RemoteMongoCollection::find(const std::string& filter_json,
+void RemoteMongoCollection::find(const bson::BsonDocument& filter_bson,
                                  RemoteFindOptions options,
-                                 std::function<void(Optional<std::string>, util::Optional<AppError>)> completion_block)
+                                 std::function<void(util::Optional<bson::BsonArray>, util::Optional<AppError>)> completion_block)
 {
     try {
-        auto base_args = m_base_operation_args;
-        base_args.push_back({"query", nlohmann::json::parse(filter_json)});
-        
+        auto base_args = bson::BsonDocument(m_base_operation_args);
+        base_args["query"] = filter_bson;
+
         if (options.limit) {
-            base_args.push_back({"limit", options.limit.value()});
+            base_args["limit"] = static_cast<int32_t>(*options.limit);
```

Comment:
Is it intentional that options.limit is sometimes int32_t and sometimes int64_t?

Reply:
No, changing.